voting/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+                let old_account_stake = self.account_stake.remove(&account_id).unwrap();
+                let account_current_stake = env::validator_stake(&account_id);
+                let account_voted_stake = self.votes.remove(&account_id).unwrap();
+                if account_current_stake > 0 {
+                    let new_account_voted_stake = (U256::from(account_voted_stake)
+                        * U256::from(account_current_stake)
+                        / U256::from(old_account_stake))
+                    .as_u128();
+                    self.total_voted_stake =
+                        self.total_voted_stake + new_account_voted_stake - account_voted_stake;
+                    self.votes.insert(&account_id, &new_account_voted_stake);
+                    self.account_stake
+                        .insert(&account_id, &account_current_stake);
+                }
+            }
+            self.check_result();
+            self.last_epoch_height = cur_epoch_height;
+        }
+    }
+
+    fn check_result(&mut self) {
+        assert!(
+            self.result.is_none(),
+            "check result is called after result is already set"
+        );
+        let total_stake = env::validator_total_stake();
+        if self.total_voted_stake > 2 * total_stake / 3 {
+            self.result = Some(env::block_timestamp());
+        }
+    }
+
+    /// Vote on a specific proposal with the given stake
+    pub fn vote(&mut self, stake: Balance) {
+        if self.result.is_some() {
+            env::panic("voting has already ended".as_bytes());
+        }
+        self.resolve_votes();
+        if self.result.is_some() {
+            return;
+        }
+        let account_id = env::predecessor_account_id();
+        let account_stake = env::validator_stake(&account_id);
+        assert!(
+            account_stake > 0,
+            "account {} is not a validator",
+            account_id
+        );
+        if stake > account_stake {
+            env::panic(
+                format!(
+                    "account {} has a stake of {} but tries to vote {}",
+                    account_id, account_stake, stake
+                )
+                .as_bytes(),
+            );
+        }
+        let voted_stake = self.votes.remove(&account_id).unwrap_or_default();
+        assert!(
+            voted_stake <= self.total_voted_stake,
+            "voted stake {} is more than total voted stake {}",
+            voted_stake,
+            self.total_voted_stake
+        );
+        if stake == 0 {
+            self.account_stake.remove(&account_id);
+            return;
+        }
+        self.votes.insert(&account_id, &stake);
+        self.account_stake.insert(&account_id, &account_stake);
+        self.total_voted_stake = self.total_voted_stake + stake - voted_stake;
+        self.check_result();
+    }
+
+    pub fn get_result(&self) -> Option<Timestamp> {
```

Comment:
hm, is Option fine for lockup and all other contracts?
should we just set a backup date that it will return or if the stake voted - it will update it to closer date?

Reply:
@evgenykuzyakov I thought it is fine to use `Option` here?

Reply:
Yes, option is expected here. I never heard about backup date?

Reply:
well, that was a suggestion if it makes it easier. We don't need it.